src/llvm/types/ir.rs

```diff
@@ -0,0 +1,339 @@
+                .collect();
+            LLVMMDNodeInContext2(
+                context,
+                elements.as_mut_slice().as_mut_ptr(),
+                elements.len(),
+            )
+        };
+        unsafe { Self::from_metadata_ref(context, metadata) }
+    }
+}
+
+pub struct MetadataEntries {
+    entries: *mut LLVMValueMetadataEntry,
+    count: usize,
+}
+
+impl MetadataEntries {
+    pub fn new(v: LLVMValueRef) -> Option<Self> {
+        if unsafe { LLVMIsAGlobalObject(v).is_null() && LLVMIsAInstruction(v).is_null() } {
+            return None;
+        }
+
+        let mut count = 0;
+        let entries = unsafe { LLVMGlobalCopyAllMetadata(v, &mut count) };
+        if entries.is_null() {
+            return None;
+        }
+
+        Some(MetadataEntries { entries, count })
+    }
+
+    pub fn iter(&self) -> impl Iterator<Item = (LLVMMetadataRef, u32)> + '_ {
+        (0..self.count).map(move |index| unsafe {
+            (
+                LLVMValueMetadataEntriesGetMetadata(self.entries, index as u32),
+                LLVMValueMetadataEntriesGetKind(self.entries, index as u32),
+            )
+        })
+    }
+}
+
+impl Drop for MetadataEntries {
+    fn drop(&mut self) {
+        unsafe {
+            LLVMDisposeValueMetadataEntries(self.entries);
+        }
+    }
+}
+
+/// Represents a metadata node.
+#[derive(Clone)]
+pub struct Function<'ctx> {
+    pub value_ref: LLVMValueRef,
+    _marker: PhantomData<&'ctx ()>,
+}
+
+impl<'ctx> Function<'ctx> {
+    /// Constructs a new [`Function`] from the given `value`.
+    ///
+    /// # Safety
+    ///
+    /// This method assumes that the provided `value` corresponds to a valid
+    /// instance of [LLVM `Function`](https://llvm.org/doxygen/classllvm_1_1Function.html).
+    /// It's the caller's responsibility to ensure this invariant, as this
+    /// method doesn't perform any valiation checks.
+    pub(crate) unsafe fn from_value_ref(value_ref: LLVMValueRef) -> Self {
+        Self {
+            value_ref,
+            _marker: PhantomData,
+        }
+    }
+
+    pub(crate) fn name(&self) -> &str {
+        symbol_name(self.value_ref)
+    }
+
+    pub(crate) fn params(&self) -> impl Iterator<Item = LLVMValueRef> {
+        let params_count = unsafe { LLVMCountParams(self.value_ref) };
+        let value = self.value_ref;
+        (0..params_count).map(move |i| unsafe { LLVMGetParam(value, i) })
+    }
+
+    pub(crate) fn basic_blocks(&self) -> impl Iterator<Item = LLVMBasicBlockRef> + '_ {
+        self.value_ref.basic_blocks_iter()
+    }
+
+    pub(crate) fn sub_program(&self, context: LLVMContextRef) -> Option<DISubprogram<'ctx>> {
+        let sub_program = unsafe { LLVMGetSubprogram(self.value_ref) };
+        NonNull::new(sub_program).map(|_| unsafe {
+            DISubprogram::from_value_ref(LLVMMetadataAsValue(context, sub_program))
+        })
+    }
+
+    pub(crate) fn set_subprogram(&mut self, sub_program: &DISubprogram) {
```

Comment:
this should be set_sub_program

Reply:
I think "subprogram" is usually written as one word together, so probably the other way around?

https://llvm.org/doxygen/classllvm_1_1DISubprogram.html
https://www.oed.com/dictionary/subprogramme_n?tab=factsheet#20089009

Reply:
I changed to "subprogram", see the commit message